testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java

```diff
@@ -2545,9 +2545,42 @@ void testCastToBoolean(CastType castType, SqlOperatorFixture f) {
     f.checkNull("cast(null as integer array) || array[1]");
   }
 
+  /** Test case for <a href="https://issues.apache.org/jira/browse/CALCITE-7063">
+   * Result type inferred for CONCAT_FUNCTION is incorrect for BINARY arguments</a>. */
+  @Test void testConcatFuncMysql() {
+    final SqlOperatorFixture f = fixture().withLibrary(SqlLibrary.MYSQL);
+
+    // test for String
+    f.setFor(SqlLibraryOperators.CONCAT_FUNCTION_MYSQL);
+    f.checkString("concat('a', x'61')", "6161", "BINARY(2) NOT NULL");
+    f.checkString("concat('a', 'b', 'c')", "abc", "VARCHAR(3) NOT NULL");
+    f.checkString("concat(cast('a' as varchar), cast('b' as varchar), "
+        + "cast('c' as varchar))", "abc", "VARCHAR NOT NULL");
+    f.checkNull("concat('a', 'b', cast(null as char(2)))");
+    f.checkNull("concat(cast(null as ANY), 'b', cast(null as char(2)))");
+    f.checkString("concat('', '', 'a')", "a", "VARCHAR(1) NOT NULL");
+    f.checkString("concat('', '', '')", "", "VARCHAR(0) NOT NULL");
+    f.checkFails("^concat()^", INVALID_ARGUMENTS_NUMBER, false);
+    f.checkString("concat('abc', 'bb', 'cc')", "abcbbcc", "VARCHAR(7) NOT NULL");
+
+    // test for ByteString
+    f.checkString("concat(x'616263',x'62')", "61626362", "BINARY(4) NOT NULL");
+    f.checkString("concat(x'616263','abc')", "616263616263", "BINARY(6) NOT NULL");
+    f.checkString("concat(x'61',x'62')", "6162", "BINARY(2) NOT NULL");
+    f.checkString("concat(cast(x'61' as binary), cast(x'62' as binary), "
+        + "cast(x'63' as binary))", "616263", "BINARY(3) NOT NULL");
+    f.checkNull("concat(x'61', x'62', cast(null as binary))");
+    f.checkNull("concat(cast(null as ANY), 'b', cast(null as binary(2)))");
+    f.checkNull("concat(cast(null as ANY), 'b', x'61')");
+    f.checkString("concat('a', x'61')", "6161", "BINARY(2) NOT NULL");
+    f.checkString("concat(x'', x'', x'61')", "61", "BINARY(1) NOT NULL");
+    f.checkString("concat(x'', x'', x'')", "", "BINARY(0) NOT NULL");
+    f.checkFails("concat(1, 'a', x'61')",
+        "concat function only accept for string or bytestring types.", true);
+  }
+
   @Test void testConcatFunc() {
     final SqlOperatorFixture f = fixture();
-    checkConcatFunc(f.withLibrary(SqlLibrary.MYSQL));
```

Comment:
I think you can keep String concat cases here, and add tests for ByteString in above. They are duplicated.

Reply:
make sense, done.